src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/CourseActivity.kt

```diff
@@ -0,0 +1,220 @@
+}
+
+@Component
+class CourseActivityBuilderFactory(
+  private val repository: CourseActivityBuilderRepository? = null,
+  private val courseActivityPayRateBuilderFactory: CourseActivityPayRateBuilderFactory = CourseActivityPayRateBuilderFactory(),
+  private val courseScheduleBuilderFactory: CourseScheduleBuilderFactory = CourseScheduleBuilderFactory(),
+  private val courseScheduleRuleBuilderFactory: CourseScheduleRuleBuilderFactory = CourseScheduleRuleBuilderFactory(),
+) {
+  fun builder(): CourseActivityBuilder {
+    return CourseActivityBuilder(
+      repository,
+      courseActivityPayRateBuilderFactory,
+      courseScheduleBuilderFactory,
+      courseScheduleRuleBuilderFactory,
+    )
+  }
+}
+
+class CourseActivityBuilder(
+  val repository: CourseActivityBuilderRepository? = null,
+  val courseActivityPayRateBuilderFactory: CourseActivityPayRateBuilderFactory,
+  val courseScheduleBuilderFactory: CourseScheduleBuilderFactory,
+  val courseScheduleRuleBuilderFactory: CourseScheduleRuleBuilderFactory,
+) : CourseActivityDsl {
+
+  private lateinit var courseActivity: CourseActivity
+
+  fun build(
+    programService: ProgramService,
+    courseActivityId: Long,
+    code: String,
+    prisonId: String,
+    description: String,
+    capacity: Int,
+    active: Boolean,
+    startDate: String,
+    endDate: String?,
+    minimumIncentiveLevelCode: String,
+    internalLocationId: Long?,
+    excludeBankHolidays: Boolean,
+  ): CourseActivity =
+    CourseActivity(
+      courseActivityId = courseActivityId,
+      code = code,
+      program = programService,
+      caseloadId = prisonId,
+      prison = lookupAgency(prisonId),
+      description = description,
+      capacity = capacity,
+      active = active,
+      scheduleStartDate = LocalDate.parse(startDate),
+      scheduleEndDate = endDate?.let { LocalDate.parse(it) },
+      iepLevel = lookupIepLevel(minimumIncentiveLevelCode),
+      internalLocation = internalLocationId?.let { lookupAgencyInternalLocation(it, prisonId) },
+      excludeBankHolidays = excludeBankHolidays,
+    )
+      .let { save(it) }
+      .also { courseActivity = it }
+
+  override fun payRate(
+    iepLevelCode: String,
+    payBandCode: String,
+    startDate: String,
+    endDate: String?,
+    halfDayRate: Double,
+  ) =
+    courseActivityPayRateBuilderFactory.builder().build(
+      courseActivity,
+      iepLevelCode,
+      payBandCode,
+      startDate,
+      endDate,
+      halfDayRate,
+    )
+      .also { courseActivity.payRates += it }
+
+  override fun courseSchedule(
+    courseScheduleId: Long,
+    scheduleDate: String,
+    startTime: String,
+    endTime: String,
+    slotCategory: SlotCategory,
+    scheduleStatus: String,
+  ) = courseScheduleBuilderFactory.builder().build(
+    courseActivity,
+    courseScheduleId,
+    scheduleDate,
+    startTime,
+    endTime,
+    slotCategory,
+    scheduleStatus,
+  )
+    .also {
+      courseActivity.courseSchedules += it
+    }
```

Comment:
An example of applying the DSL to one of CourseActivity's children. Here we:
* get a builder from the child's builder factory
* build the child, which saves itself, and return the entity so that it can be assigned to a variable in the test
* add the child to the CourseActivity so we don't have to reload the CourseActivity from the db